developer_docs/how_to_create_a_plugin.asciidoc

```diff
@@ -0,0 +1,2137 @@
+----
+my_host.refresh_statuses
+----
+The method refreshes *all* statuses by default, this is usually not what
+you want so provide status for refresh.
+
+[source, ruby]
+----
+my_host.refresh_statuses(HostStatus.find_status_by_humanized_name("statusname")
+----
+[[selecting-properties-to-clone]]
+==== Selecting properties to clone
+
+_Requires Foreman 1.11 or higher, set `requires_foreman '>= 1.11'` in
+engine.rb_
+
+If you extend the Host::Managed object and add attributes or
+associations to the model, you probably want those to be cloned with the
+rest of the host object. +
+In your concern you should add the following calls:
+
+[source, ruby]
+----
+module ForemanPluginTemplate
+  module HostExtensions
+    extend ActiveSupport::Concern
+
+    included do
+      # specify which properties to include in clone
+      include_in_clone :property1, :property2
+
+      # specify which properties should not be cloned
+      exclude_from_clone :property3, :property4
+    end
+  end
+end
+----
+All attributes on the model will be cloned by default (therefore may be
+excluded), while associations to other models will _not_ be cloned by
+default (therefore may be included).
+
+[[host-info-providers]]
+==== Host info providers
+
+Every host exposes `Host#info` method to provide a complete information
+hash about itself. This hash is mainly used as external node classifier
+in puppet. +
+Any plugin can extend this info by creating a class that inherits
+`HostInfo::Provider` and registering it in the plugin:
+
+[source]
+----
+# In plugin declaration (engine.rb):
+Foreman::Plugin.register :my_plugin do
+  register_info_provider MyPlugin::InfoProvider
+end
+
+# Actual info provider class
+module MyPlugin
+  class InfoProvider < HostInfo::Provider # inherit the base class
+
+    # override this method according to principles specified below
+    def host_info
+      { 'parameters' => host.params }
+    end
+  end
+end
+----
+Info hash is structured in the following way:
+
+[source, ruby]
+----
+host_info = Host.first.info
+
+host_info['classes'] # set of puppet classes that are associated with this host including class parameters
+host_info['parameters'] # list of foreman properties that are associated with this host i.e taxonomy, hostgroup, interfaces.
+# This list also includes values of global parameters associated with the host.
+host_info['environment'] # Host's environment
+
+----
+[[settings]]
+=== Settings
+
+Plugins can store Foreman-wide settings either in the database or a
+config file. The DB should be preferred as it can be managed from the UI
+(under Administer > Settings), CLI, API and Puppet, and changed on the
+fly, while the config file is usually only used for settings that change
+behaviour during app startup and require a restart.
+
+To add new DB settings, the plugin should define a Setting class named
+after the plugin. For foreman_example, this would be at
+`app/models/setting/example.rb` and should look like:
+
+....
```

Comment:
no colors for ruby source :(, same in the 2 following examples with `initializer`